sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SizeBasedCoalescer.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.Partition
+import org.apache.spark.rdd._
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.logical.statsEstimation.EstimationUtils
+
+class SizeBasedCoalescer(val maxSize: Long, val schema: Seq[Attribute])
+  extends PartitionCoalescer with Serializable {
+  override def coalesce(maxPartitions: Int, parent: RDD[_]): Array[PartitionGroup] = {
+    val rowSize = EstimationUtils.getSizePerRow(schema).toLong
+    val partitionIndexToSize = parent.mapPartitionsWithIndexInternal((index, part) => {
+      // TODO make it more accurate
+      Map(index -> rowSize * part.size).iterator
+    }).collectAsMap()
```

Comment:
Is it too costly to trigger an action to compute all the RDDs?

Reply:
> Is it too costly to trigger an action to compute all the RDDs?

this is after compute rdds and coalese computed rdd.

Reply:
> Is it too costly to trigger an action to compute all the RDDs?

Sorry for I have make some mistake, it will re-compute last stage. but won't recompute all stage